bc_obps/service/form_builder_service.py

```diff
@@ -0,0 +1,196 @@
+        st_schema['properties']['emissions']['items']['dependencies'] = gas_type_one_of
+    # Add the source_type schema to the schema object being returned by this function
+    rjsf_schema['properties']['sourceTypes']['properties'][
+        str_to_camel_case(source_type_schema.source_type.name)
+    ] = st_schema
+
+
+# build_schema will dynamically create a form depending on the parameters passed
+# activity: Returns a form with just the activity schema + the set of related source_types
+# activity + source_type(s): Returns the activity schema, plus the schema(s) for each source type id passed in the List, plus fills the gas_type enum with the valid gas types based on the activity & source type
+# activity + source_type(s) + gas_type selection is made: Returns all of the above, plus fills out the methodology enum with the valid methodology options based on the activity, source_type & gas_type selected
+# activity + source_type(s) + gas_type selection + methodology selection is made: Returns all of the above, plus the additional reporting fields associated to the methodology selection
+# report_date is mandatory & determines the valid schemas & WCI configuration for the point in time that the report was created
+
+
+def build_schema(activity: int, source_types: List[int], report_date: str):
+    # Get activity schema
+    rjsf_schema = (
+        ActivityJsonSchema.objects.filter(
+            reporting_activity_id=activity, valid_from__valid_from__lte=report_date, valid_to__valid_to__gte=report_date
+        )
+        .first()
+        .json_schema
+    )
+    gas_type_map = {}
+    methodology_map = {}
+
+    # Fetch valid source_type(s) for the activity
+    valid_source_types = (
+        ConfigurationElement.objects.select_related('source_type')
+        .filter(
+            reporting_activity_id=activity, valid_from__valid_from__lte=report_date, valid_to__valid_to__gte=report_date
+        )
+        .distinct('source_type__id')
+    )
+
+    # Except if no valid source_types are found
+    if valid_source_types.count() == 0:
+        raise Exception(f'No valid source_types found for activity_id {activity} & report_date {report_date}')
+    # If an activity only has one source_type, the source type is mandatory and should be added to the schema
+    elif valid_source_types.count() == 1:
+        rjsf_schema['properties']['sourceTypes'] = {"type": "object", "title": "Source Types", "properties": {}}
+        build_source_type_schema(
+            activity, valid_source_types[0].source_type_id, report_date, rjsf_schema, gas_type_map, methodology_map
+        )
+
+    # If there are multiple source_types for an activity, the user may choose which ones apply. The IDs of the selected source_types are passed as a list in the parameters & we add those schemas to the activity schema.
+    else:
+        for s in valid_source_types:
+            rjsf_schema['properties'][str_to_camel_case(s.source_type.name)] = {
```

Comment:
consider adding a jsonKey column on the source_type table so that we don't have to rely on the camel_case function to generate our keys from a string name

Reply:
jsonkey is not going to work here some of the source types have incredibly long names. Suggest exploring using IDs as keys

Reply:
or a shortened jsonkey